src/efftester.ml

```diff
@@ -1345,6 +1568,34 @@ let rec tcheck env term =
        then (et, no_eff)
        else failwith "tcheck: variable types disagree"
      with Not_found -> failwith "tcheck: unknown variable")
+  | ListTrm (typ, lst, eff) ->
+    (match typ with
+    | List elem_typ ->
+      List.iter
+        (fun e ->
+          if not (imm_type e = elem_typ)
+          then failwith "tcheck: a list type mismatches its element's type")
+        lst;
+      (typ, eff)
+    | _ -> failwith "tcheck: ListTrm must have a list type")
+  | Constructor (typ, name, payload_lst, eff) ->
+    (match check_opt_invariants (typ, name, payload_lst) with
+    | Ok _ ->
+      (try
+         List.iter (fun trm -> tcheck env trm |> ignore) payload_lst;
+         (typ, eff)
+       with Failure msg -> failwith msg)
+    | Error msg -> failwith msg)
+  | PatternMatch (typ, _matched_trm, cases, eff) ->
+    let has_type_mismatch typ1 typ2 = if types_compat typ1 typ2 then false else true in
+    let has_type_mismatch_lst =
+      List.exists (fun (_pat, body) -> has_type_mismatch typ (imm_type body)) cases
+    in
+    (* FIXME: it fails to typecheck with seed 430633514
+      Debugging using printing shows that types are same but it still fails *)
```

Comment:
I can't reproduce the failure on my machine with this seed. Are you running `ci_tests.exe` or `effmain.exe`, can you reproduce the issue from the current state o the branch?

Reply:
Ah, that is reproduced only when there is `type_check_test` is in the `ci_tests`.

Reply:
I debugged it by printing the types when they are not compatible according `types_compat` fn. However, when printed, the types are the same.